testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/CIBATest.java

```diff
@@ -0,0 +1,1196 @@
+        try (CloseableHttpResponse res = oauth.doLogout(refreshToken, TEST_CLIENT_PASSWORD)) {
+            assertThat(res, Matchers.statusCodeIsHC(Status.NO_CONTENT));
+        }
+
+        // confirm logged out
+        OAuthClient.AccessTokenResponse tokenRes = oauth.doRefreshTokenRequest(refreshToken, TEST_CLIENT_PASSWORD);
+        Assert.assertThat(tokenRes.getStatusCode(), is(equalTo(400)));
+        Assert.assertThat(tokenRes.getError(), is(equalTo(OAuthErrorException.INVALID_GRANT)));
+        if (isOfflineAccess) Assert.assertThat(tokenRes.getErrorDescription(), is(equalTo("Offline user session not found")));
+        else Assert.assertThat(tokenRes.getErrorDescription(), is(equalTo("Session not active")));
+
+        return events.expectLogout(sessionId).client(TEST_CLIENT_NAME).user(AssertEvents.isUUID()).session(AssertEvents.isUUID()).clearDetails().assertEvent();
+    }
+
+    private EventRepresentation doTokenRevokeByRefreshToken(String refreshToken, String sessionId, String userId, boolean isOfflineAccess) throws IOException{
+        try (CloseableHttpResponse res = oauth.doTokenRevoke(refreshToken, "refresh_token", TEST_CLIENT_PASSWORD)) {
+            assertThat(res, Matchers.statusCodeIsHC(Status.OK));
+        }
+
+        // confirm revocation
+        OAuthClient.AccessTokenResponse tokenRes = oauth.doRefreshTokenRequest(refreshToken, TEST_CLIENT_PASSWORD);
+        Assert.assertThat(tokenRes.getStatusCode(), is(equalTo(400)));
+        Assert.assertThat(tokenRes.getError(), is(equalTo(OAuthErrorException.INVALID_GRANT)));
+        if (isOfflineAccess) Assert.assertThat(tokenRes.getErrorDescription(), is(equalTo("Offline user session not found")));
+        else Assert.assertThat(tokenRes.getErrorDescription(), is(equalTo("Session not active")));
+
+        return events.expect(EventType.REVOKE_GRANT).clearDetails().client(TEST_CLIENT_NAME).user(AssertEvents.isUUID()).assertEvent();
+    }
+
+    private void testBackchannelAuthenticationFlow(boolean isOfflineAccess) throws Exception {
+        ClientResource clientResource = null;
+        ClientRepresentation clientRep = null;
+        try {
+            final String username = "nutzername-rot";
+            final String bindingMessage = "BASTION";
+
+            // prepare CIBA settings
+            clientResource = ApiUtil.findClientByClientId(adminClient.realm(TEST_REALM_NAME), TEST_CLIENT_NAME);
+            clientRep = clientResource.toRepresentation();
+            prepareCIBASettings(clientResource, clientRep);
+            if(isOfflineAccess) oauth.scope(OAuth2Constants.OFFLINE_ACCESS);
+
+            // user Backchannel Authentication Request
+            AuthenticationRequestAcknowledgement response = doBackchannelAuthenticationRequest(TEST_CLIENT_NAME, TEST_CLIENT_PASSWORD, username, bindingMessage);
+
+            // user Authentication Channel Request
+            AuthenticationChannelRequest authenticationChannelReq = doAuthenticationChannelRequest();
```

Comment:
@tnorimat  @pedroigor This is only testsuite related concern, but my vote is that method "doAuthenticationChannelRequest()" accepts the bindingMessage as an argument and that the bindingMessage is sent to the testing endpoint. And also that the testing endpoint checks the "bindingMessage" equals to the previously saved request. This is minor testsuite related concern, but IMO it may help the "newbie" to understand CIBA protocols a bit better.

In general, the CIBA protocol is quite hard to understand for the newbies IMO as there is big amount of various HTTP requests sent among various parties. And hence it helps a bit if the testsuite related requests are also a bit "logical" and represents the reality, which will be used by the CIBA protocol in production. And IMO in reality, the request from the user to AuthenticationChannel will need to contain the "bindingMessage" as in reality AuthenticationChannel can contain many different requests from different users and bindingMessage would be used to bind the request with the actual user. For same reason, I vote to change the BlockingQueue in the TestingOIDCEndpointsApplicationResource to be the Map where the key is the "bindingMessage" and the value the actual request. WDYT?

Reply:
@mposolda I agree with you. It is quite reasonable. I will revise `CIBATest` as you suggested.

Reply:
Thanks!